### ISSUE TITLE
Getting NaNs in the pullback of ReverseSequence

### ISSUE THREAD

**Issue by allioux** [LuxDL/Lux.jl#1043]
Hi, I was playing with a model using BidirectionalRNN and I was getting NaNs when computing its gradient. I then isolated the following problematic model involving two ReverseSequence layers:

```julia
using Lux
using Zygote
using Random
using Optimisers

function Model()
    rnn = Recurrence(RNNCell((30 => 30)); return_sequence=true)
    Chain(ReverseSequence(), rnn, ReverseSequence(), x -> x[end])
end

rng = Random.default_rng()
Random.seed!(rng, 0)
model = Model()
model = Lux.Experimental.@debug_mode model nan_check = :both
ps, st = Lux.setup(rng, model)
opt = Adam()
tstate = Training.TrainState(model, ps, st, opt)
vjp_rule = AutoZygote()
loss_function = CrossEntropyLoss(; logits=Val(true))

x = rand(Float64, (30, 2, 5))
y = rand(Float64, (30, 5))

try
    (loss, _, _), back = Zygote.pullback(loss_function, tstate.model, tstate.parameters, tstate.states, (x, y))
    grads = back((one(loss), nothing, nothing))[2]
catch e
    println(e)
end
```

The reported error is the following:
> NaNs detected in pullback output (x)  of layer ReverseSequence{Nothing}(nothing) at location KeyPath(:model, :layers, :layer_1).

I wasn't able to simplify this model much further. Similarly, if I reduce the dimensions involved (e.g., using a RNNCell(10 => 10) instead of a RNNCell(30 => 30) ), NaNs are no longer reported.

**Comment by avik-pal** [comment]
@allioux thanks for catching this. We weren't zeroing out the array created by `similar` which caused this. I will merge that patch once tests pass

**Comment by allioux** [comment]
Thank you for this quick reply!
